_posts/2020-04-01-#18401.md

```diff
@@ -0,0 +1,133 @@
+  explanations.
+
+- To resolve the quadratic hashing problem, the way that the transaction digest
+  is calculated was changed for segwit inputs. See [BIP
+  143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) for the
+  motivation and specification.
+
+- The segwit transaction digest is designed so parts of the digest are shared
+  between all the signatures in a transaction. This means that the total amount
+  of data to be hashed for the transaction is linear in the size of the
+  transaction.
+
+- This shared data between all signatures of a transaction is stored in the
+  `PrecomputedTransactionData` object, which was added in [PR
+  8524](https://github.com/bitcoin/bitcoin/pull/8524).
+
+- Segwit v1 (taproot) makes some very minor changes to the transaction
+  digest algorithm. Specifically, the transaction digest commits to the
+  amounts of all inputs (instead of just the amount of the input being signed)
+  and to the scriptPubKey of the output being spent. See [BIP
+  341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message)
+  for the full description of the transaction digest.
+
+- Because of these changes, more data needs to be stored in the
+  `PrecomputedTransactionData` object. This PR changes the way that the object
+  is constructed and initialized so that a [future
+  commit](https://github.com/bitcoin/bitcoin/pull/17977/commits/dd0622b8e2f8a324fdf73eb476f136ee556f3e68)
+  can add that data to the object.
+
+### Move single-sig checking EvalScript code to EvalChecksig
+
+- [Tapscript](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki)
+  is the scripting language used for scripts in a taproot tree (taproot has
+  versioning for scripts, so a future soft fork could introduce a different
+  scripting version or language).
+
+- Tapscript has almost the same semantics as Bitcoin script. One notable
+  difference is that `OP_CHECKSIG` and `OP_CHECKSIGVERIFY` both verify schnorr
+  signatures instead of ECDSA signatures.
+
+- This PR extracts most of the logic for `OP_CHECKSIG` and
+  `OP_CHECKSIGVEIFY` from the very large `EvalScript()` function.
+
+- A [future
+  commit](https://github.com/bitcoin/bitcoin/pull/17977/commits/af411a6cbec12375627aff374874a7f24b7c07e9)
+  modifies this function to switch on which flags have been passed into the
+  script interpreter in order to determine whether to do schnorr or ECDSA
+  signature verification.
+
+## Questions
+
+1. Did you review the PRs? [Concept ACK, approach ACK, tested ACK, or
+   NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)
+   (Don't forget to put your PR review on GitHub.)
+
+2. A new `PrecomputedTransactionData` default constructor is added which
+   doesn't initialize the data members. Where do they get initialized instead?
+
+3. What new data members are added to `PrecomputedTransationData` in the
+   subsequent commit in the taproot PR?
```

Comment:
maybe specify the PR 17977, or better yet, the commit (as there are 17 of them in 17977)

Reply:
Added the PR number, but left finding the exact commit as a (very easy - it's in the commit log summary) exercise for the reader.